### ISSUE TITLE
Contradicting switch with return

### ISSUE THREAD

**Issue by mkroeders** [slevomat/coding-standard#809]
Currently there there are two contradicting rules. `return` requires no enters after it, while `case` requires enters after it. Both rules I want to keep but in this instance I want to use the `case` rule so it is more clear of the separate cases. unforunatly I cannot see a parameter option of any kind to do this.

Code

The following code states (correctly) that there is an enter missing

    switch ($baz) {
        case 'foo':
            return 'foo';
        case 'bar':
            return 'bar';
    }

(phpcs output)

    136 | ERROR | [x] Expected 1 lines after "case", found 0.
    (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountAfterControlStructure)
    138 | ERROR | [x] Expected 1 lines before "case", found 0.
    (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountBeforeControlStructure)

To fix this an enter is added between the two cases, however this triggers the an issue with an enter after the `return`

    switch ($orderType) {
        case 'foo':
            return 'foo';

        case 'bar':
            return 'bar';
    }

(phpcs output)

    137 | ERROR | [x] Expected 0 lines after "return", found 1.
    (SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing.IncorrectLinesCountAfterLastControlStructure)

I am I missing an option that can fix this or is it one or the other rule?

The [phpcs config](https://gist.github.com/mkroeders/98b5f9e56e5e0f1e2c3b53a1e468c71a) in use

**Comment by kukulich** [comment]
You can disable sniffs for specific tokens.

**Comment by mkroeders** [comment]
@kukulich how do you do that? I am looking at the sniff for the return (https://github.com/slevomat/coding-standard/blob/master/SlevomatCodingStandard/Sniffs/ControlStructures/JumpStatementsSpacingSniff.php#L45). But do not see the option. Normally you see parameters with which you can do that.

Or do you mean to disable using comments?

**Comment by kukulich** [comment]
https://github.com/slevomat/coding-standard/blob/master/build/phpcs.xml

**Comment by mkroeders** [comment]
@kukulich thanks, I copied the config for `JumpStatementsSpacing` which will allow the code above (switch setup). However then the following code will also pass (an extra enter after the return), which was previously not allowed which is what I want.

    public function test(): ?int
    {
        return null;

    }

So it is specific that a return inside a switch should allow the enter.

**Comment by mkroeders** [comment]
@kukulich I just put the code below in a function and the whole thing blew up

    <?php
    $var = 'foo';
    switch ($var) {
        case 'foo':
            echo 'foo';
            break;
    
        case 'bar':
            echo 'bar';
            break;
    }

```
Uncaught TypeError: Argument 2 passed to SlevomatCodingStandard\Sniffs\Classes\PropertySpacingSniff::isNextMemberValid() must be of the type int, null given, called in /vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Classes/AbstractPropertyAndConstantSpacing.php on line 68 and defined in /vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Classes/PropertySpacingSniff.php:46
Stack trace:
#0 /vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Classes/AbstractPropertyAndConstantSpacing.php(68): SlevomatCodingStandard\Sniffs\Classes\PropertySpacingSniff->isNextMemberValid(Object(PHP_CodeSniffer\Files\LocalFile), NULL)
#1 /vendor/slevomat/coding-standard/SlevomatCodingStandard/Sniffs/Classes/PropertySpacingSniff.php(43): SlevomatCodingStandard\Sniffs\Classes\AbstractPropertyAndConstantSpac in /vendor/slevomat/coding-standard/SlevomatCoding
Standard/Sniffs/Classes/PropertySpacingSniff.php on line 46
```

To test it, i went back to 6.1 and there it also breaks. So not sure what is wrong

**Comment by kukulich** [comment]
@mkroeders I see no error: https://github.com/slevomat/coding-standard/commit/d6d032a2da97bedfe57bbcaaed756dc7321cb54d

Please post the whole code that throws this error.

**Comment by mkroeders** [comment]
@kukulich I think i found the issue. (tested using 6.1.5)

The issue comes from `PropertySpacingSniff` which extends `AbstractPropertyAndConstantSpacing`.

On line 64 the following code is done

```php
$types = [T_COMMENT, T_DOC_COMMENT_OPEN_TAG, T_CONST, T_VAR, T_PUBLIC, T_PROTECTED, T_PRIVATE];
$nextPointer = TokenHelper::findNext($phpcsFile, $types, $firstOnLinePointer + 1);
```

My understanding is that it obtains the next token after `$types` tokens. When looking into the code where it goes wrong it ends up on the following piece;

```php
$mock = new class($container) {
    use ResourceServiceContainerAware {
        getResourceServiceByNameFromContainer as public;
        getResourceServiceContainer as public;
    }
};
```

This piece makes a mock for `ResourceServiceContainerAware` and makes the `getResourceServiceByNameFromContainer` & `getResourceServiceContainer` into public methods.

However looking in the lint code it probably matches the public token. However this case it should not match.

**Comment by mkroeders** [comment]
As said upgraded to `dev-master` I got the error also. So I added the following piece to the PropertySpacingSniff::process and it is solved

```php
        $asPoint = TokenHelper::findPrevious($phpcsFile, [T_AS], $pointer);
        if ($asPoint !== null) {
            return $pointer;
        }
```

What it does, it checks if there is a "as" before. I however don't know if this is the solution to use, but this makes it possible to test the next step the switch

**Comment by mkroeders** [comment]
I don't know if you prefer a new comment or updating the previous one. For now I do a new one.

WIth the previously created fix I started the switch testing.

```php
<?php
declare(strict_types=1);

/**
 * Tests switch linter
 *
 * @param string $foo
 *
 * @return string
 */
public function testSwitch(string $foo): string
{
    switch ($foo) {
        case 'foo':
            return 'foo';

        case 'bar':
            return 'bar';

        default:
            return 'baz';
    }
}
```

Is my desired code. When runnning this I got

```
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 15 | ERROR | [x] Expected 0 lines after "return", found 1.
 18 | ERROR | [x] Expected 0 lines after "return", found 1.
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

Fixing the above errors leads to

```
----------------------------------------------------------------------
FOUND 4 ERRORS AFFECTING 3 LINES
----------------------------------------------------------------------
 14 | ERROR | [x] Expected 1 lines after "case", found 0.
 16 | ERROR | [x] Expected 1 lines before "case", found 0.
 16 | ERROR | [x] Expected 1 lines after "case", found 0.
 18 | ERROR | [x] Expected 1 lines before "default", found 0.
----------------------------------------------------------------------
PHPCBF CAN FIX THE 4 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

After disabling all my "changes" to the rules it seems to stay, so unfortunately it seems not be resolved

**Comment by kukulich** [comment]
@mkroeders the errors looks right. It looks you have incompatible configuration in two sniffs.

**Comment by mkroeders** [comment]
@kukulich I looked at 34db1f1, and saw the new options. And yes, that fixes my issue :) Thanks for all the help

Needed to add the following to my rules

```xml
<rule ref="SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing">
    <properties>
        <property name="linesCountAfterWhenLastInCaseOrDefault" value="1"/>
    </properties>
</rule>
```

**Comment by genesiscz** [comment]
> @mkroeders the errors looks right. It looks you have incompatible configuration in two sniffs.

I also have the same problem with 6.2.0

My case:

```php
		switch($badge) {
			case "record_all_time":
				// we have to delete the previous record
				doStuff()
				break;
			case "record_month":
				//we have to delete records from this month first
				doOtherStuff()
				break;
			case "record_day":
                               anotherStuff();
				break;
		}

```

Getting

>   405 | ERROR | [x] Expected 1 lines before "switch", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountB
> eforeControlStructure)
>  406 | ERROR | [x] Expected 1 lines after "case", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountA
> fterControlStructure)
>  410 | ERROR | [x] Expected 1 lines before "break", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing.IncorrectLinesCountBeforeCo
> ntrolStructure)
>  411 | ERROR | [x] Expected 1 lines before "case", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountB
> eforeControlStructure)
>  411 | ERROR | [x] Expected 1 lines after "case", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountA
> fterControlStructure)
>  417 | ERROR | [x] Expected 1 lines before "break", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing.IncorrectLinesCountBeforeCo
> ntrolStructure)
>  418 | ERROR | [x] Expected 1 lines before "case", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.BlockControlStructureSpacing.IncorrectLinesCountB
> eforeControlStructure)
>  423 | ERROR | [x] Expected 1 lines before "break", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing.IncorrectLinesCountBeforeCo
> ntrolStructure)
>  434 | ERROR | [x] Expected 1 lines before "return", found 0.
>      |       |     (SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing.IncorrectLinesCountBeforeCo
> ntrolStructure)


My ruleset file just excludes a couple of rules, but not setting any. Shouldn't it be, by default, compatible?

**Comment by genesiscz** [comment]
Toi be honest I am not quite sure we're talking about the same issue here, but I actually tried to set up the properties. Using dev-master just to try. 

So I tried

 ```xml
       <rule ref="SlevomatCodingStandard.ControlStructures.JumpStatementsSpacing">
            <properties>
                <property name="linesCountBeforeWhenFirstInCaseOrDefault" value="1"/>
                <property name="linesCountAfterWhenLastInCaseOrDefault" value="1"/>
                <property name="linesCountAfterWhenLastInLastCaseOrDefault" value="1" />
            </properties>
        </rule>
```

This has been placed into `<rule ref="vendor/slevomat/...."></rule>` 

I am not actually sure the sniffer gives me the correct lines. When I use 

`vendor/bin/phpcbf --standard=Speeder app/model/
Results.php -vv -sp`

some of these passes look like this

`=> Fixing file: 6/148 violations remaining [made 16 passes]...  * fixed 6 violations, starting loop 17
*
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:236
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:239 replaced token 2781 (T_
WHITESPACE on line 450) "\n············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n\n"
                **** SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 has possible conflict
 with another sniff on loop 15; caused by the following change ****
                **** replaced token 2781 (T_WHITESPACE on line 450) "\n············" => "············" ****
                **** ignoring all changes until next loop ****
        => Changeset failed to apply
        => Fixing file: 0/148 violations remaining [made 17 passes]...  * fixed 0 violations, starting loop 18
*
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:236
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:239 replaced token 2781 (T_
WHITESPACE on line 450) "\n············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n\n"
                A: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 replaced token 2781 (T_
WHITESPACE on line 450) "\n············" => "············"
                A: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n\n"
        => Changeset ended: 2 changes applied
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:170
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:177 replaced token 2781 (T_
WHITESPACE on line 450) "············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:182 replaced token 2780 (T_
SEMICOLON on line 450) ";\n\n" => ";\n\n\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:182 replaced token 2780 (T_
SEMICOLON on line 450) ";\n\n" => ";\n\n\n\n"
                * token 2781 has already been modified, skipping *
        => Changeset failed to apply
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:236
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:239 replaced token 2883 (T_
WHITESPACE on line 458) "\n············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2882 (T_
SEMICOLON on line 458) ";" => ";\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2882 (T_
SEMICOLON on line 458) ";" => ";\n\n"
                A: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 replaced token 2883 (T_
WHITESPACE on line 458) "\n············" => "············"
                A: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 replaced token 2882 (T_
SEMICOLON on line 458) ";" => ";\n\n"
        => Changeset ended: 2 changes applied
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:170
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:177 replaced token 2883 (T_
WHITESPACE on line 458) "············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:182 replaced token 2882 (T_
SEMICOLON on line 458) ";\n\n" => ";\n\n\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:182 replaced token 2882 (T_
SEMICOLON on line 458) ";\n\n" => ";\n\n\n\n"
                * token 2883 has already been modified, skipping *
        => Changeset failed to apply
        => Fixing file: 4/148 violations remaining [made 18 passes]...  * fixed 4 violations, starting loop 19
*
        => Changeset started by SlevomatCodingStandard.ControlStructures.AbstractControlStructur:236
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:239 replaced token 2781 (T_
WHITESPACE on line 450) "\n\n" => "\n"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:239 replaced token 2782 (T_
WHITESPACE on line 451) "\n············" => "············"
                Q: SlevomatCodingStandard.ControlStructures.AbstractControlStructur:248 replaced token 2780 (T_
SEMICOLON on line 450) ";" => ";\n"
                **** SlevomatCodingStandard.ControlStructures.AbstractControlStructur:252 has possible conflict
 with another sniff on loop 17; caused by the following change ****
                **** replaced token 2781 (T_WHITESPACE on line 450) "\n\n" => "\n" ****
                **** ignoring all changes until next loop ****
        => Changeset failed to apply
        => Fixing file: 0/148 violations remaining [made 19 passes]...  * fixed 0 violations, starting loop 20
*`

are these lines "on line 450" actually corresponding with source, or is it giving some possible modifications into the mix?

If the first one is the case, here's the screen of actual code:
![image](https://user-images.githubusercontent.com/1512353/78462917-33f59480-76d7-11ea-8e67-f8591f562bc4.png).

It kinda actually doesn't make sense that it's the line from the file since line 450 doesn't have semicolon at all.

I'd love to help you more determine what's the cause of this, but I need to poke a little bit about what exactly you need to be able to debug this.

**Comment by genesiscz** [comment]
Setting
`<property name="linesCountAfterWhenLastInCaseOrDefault" value="1"/>` fixed the inifnite loop issue, but doesn't work as expected, I suppose. The code it generated looks weirdly indented:

```php
       switch ($badge) {
            case "record_all_time":
                dostuff()

                break;

                case "record_month":

                dostuff()

                break;

                case "record_day":

                dostuff()

                break;
        }

```

**Comment by mkroeders** [comment]
@genesiscz after some testing the setting `<property name="linesCountAfterWhenLastInLastCaseOrDefault" value="0"/>` needs to be set

When looking in the code the defaults are null for both. So that is way it is not working when not setting the rules. So that is "expected" behavior.

I would recommended creating a separate issue and linking this one about the indenting, because am not sure it is from this and due to the fact this one is also closed
